### ISSUE TITLE
String.intern use in SIPHeader is quite CPU intensive

### ISSUE THREAD

**Issue by TristanPerry** [usnistgov/jsip#55]
From profiling the jsip project, I have noticed that the String.intern use within SIPHeader's constructor is fairly CPU intensive - taking up 26%+ of the overall CPU time:

![jsip-intern-1](https://user-images.githubusercontent.com/8782943/72528818-dda95000-3863-11ea-9a9e-fc3cdae5b803.png)

![jsip-intern-2](https://user-images.githubusercontent.com/8782943/72528826-e26e0400-3863-11ea-85df-2326395ed865.png)

This [has been noticed](https://github.com/RestComm/jain-sip/issues/13) in other sip stack forks, albeit an attempted fix (moving to a hashmap) was reverted for an unknown reason.

Since there's only ever a handful of header names, and they are never compared with `==`, is `.intern()` fully needed - or is there a better, less CPU intensive way of handling this?

Apologies if this has already been discussed and there's no desire to change it; I done some searching and couldn't see any mention, so thought I'd raise this issue incase it wasn't known.

**Comment by vladimirralev** [comment]
Yes this is a known pattern and a number of things have been tried over the years. In general profiling a JVM for CPU in apps that are not CPU-bound is not worth it. Most jsip apps I've seen are memory bound. I remember avoiding intern() had an adverse effect on garbage collection. You may gain some speed until the first major GC event hits you. And garbage collection is not subject to a Java profiler so it gives inaccurate results.

Not familiar with the results of this particular patch here but it's possible that without synchronisation it had some issues, while with synchro it would be similar to intern().

There are new garbage collectors now and it may be worth re-examining the issue, however I wouldn't trust a JVM profiler to conclude this is a bottleneck, this needs a native profiler to capture GC codepaths under sustained load. You may be trading CPU for GC stress.

**Comment by TristanPerry** [comment]
Thanks for the quick reply, that's all interesting.

I didn't appear to see major GC trade-offs when performing the same tests without the `SIPHeader` `.intern()`, but that was still with Jprofiler which only shows general GC activity. (And a JVM running with the G1 GC, FWIT).

Interesting point about using a native profiler to get a better GC picture, thanks.
